File: java-quarkus-resteasy/app/src/main/java/science/monke/incoming/TaskApi.java
package science.monke.incoming;

import jakarta.enterprise.inject.Instance;
import jakarta.inject.Inject;
import jakarta.ws.rs.*;
import jakarta.ws.rs.core.MediaType;
import jakarta.ws.rs.core.Response;
import java.util.UUID;
import java.util.stream.Collectors;
import science.monke.outgoing.TaskRepoPort;

@Path("/tasks")
public class TaskApi {

  private final Instance<TaskRepoPort> taskRepoPort;

  @Inject
  public TaskApi(final Instance<TaskRepoPort> taskRepoPort) {
    this.taskRepoPort = taskRepoPort;
  }

  @GET
  @Produces(MediaType.APPLICATION_JSON)
  public Response getTasks() {
    return Response.ok(
            taskRepoPort.get().findAll().stream()
                .map(TaskResponse::fromTask)
                .collect(Collectors.toList()))
        .build();
  }

  @POST
  @Consumes(MediaType.APPLICATION_JSON)
  @Produces(MediaType.APPLICATION_JSON)
  public Response postTask(final TaskRequest taskRequest) {
    return Response.status(Response.Status.CREATED).entity(
            TaskResponse.fromTask(taskRepoPort.get().save(TaskRequest.toTask(taskRequest))))
        .build();
  }

  @GET
  @Path("/{taskId}")
  @Produces(MediaType.APPLICATION_JSON)
  public Response getTask(@PathParam("taskId") final UUID taskId) {
    return taskRepoPort
        .get()
        .findByTaskId(taskId)
        .map(task -> Response.ok(TaskResponse.fromTask(task)).build())
        .orElse(Response.status(Response.Status.NOT_FOUND).build());
  }

  @PUT
  @Path("/{taskId}")
  @Consumes(MediaType.APPLICATION_JSON)
  @Produces(MediaType.APPLICATION_JSON)
  public Response putTask(@PathParam("taskId") final UUID taskId, final TaskRequest taskRequest) {
    return taskRepoPort
        .get()
        .update(TaskRequest.toTask(taskId, taskRequest))
        .map(task -> Response.noContent().build())
        .orElse(Response.status(Response.Status.NOT_FOUND).build());
  }

  @DELETE
  @Path("/{taskId}")
  @Produces(MediaType.APPLICATION_JSON)
  public Response deleteTask(@PathParam("taskId") final UUID taskId) {
    return taskRepoPort.get().delete(taskId)
        ? Response.noContent().build()
        : Response.status(Response.Status.NOT_FOUND).build();
  }
}

File: rust-axum/app/src/incoming/task_api.rs
use crate::outgoing::task_repo::DynTaskRepository;
use axum::{
    extract::{Path, State},
    http::StatusCode,
    Json,
};

use super::models::{new_task_response_from_task_entity, TaskRequest, TaskResponse};

pub async fn get_tasks(
    State(task_repository): State<DynTaskRepository>,
) -> (StatusCode, Json<Vec<TaskResponse>>) {
    let task_entities = task_repository.lock().await.find_all().await;
    return (
        StatusCode::OK,
        Json(
            task_entities
                .into_iter()
                .map(new_task_response_from_task_entity)
                .collect(),
        ),
    );
}

pub async fn post_tasks(
    State(task_repository): State<DynTaskRepository>,
    Json(task_request): Json<TaskRequest>,
) -> (StatusCode, Json<TaskResponse>) {
    let task_entity = task_repository
        .lock()
        .await
        .save(task_request.to_new_task_entity())
        .await;
    return (
        StatusCode::CREATED,
        Json(new_task_response_from_task_entity(task_entity)),
    );
}

pub async fn get_task(
    State(task_repository): State<DynTaskRepository>,
    Path(task_id): Path<uuid::Uuid>,
) -> (StatusCode, Json<Option<TaskResponse>>) {
    let task_entity_option = task_repository.lock().await.find_by_task_id(task_id).await;
    return match task_entity_option {
        Some(task_entity) => (
            StatusCode::OK,
            Json(Some(new_task_response_from_task_entity(task_entity))),
        ),
        None => (StatusCode::NOT_FOUND, Json(None)),
    };
}

pub async fn put_task(
    State(task_repository): State<DynTaskRepository>,
    Path(task_id): Path<uuid::Uuid>,
    Json(task_request): Json<TaskRequest>,
) -> StatusCode {
    let task_entity_option = task_repository
        .lock()
        .await
        .update(task_request.to_updated_task_entity(task_id))
        .await;
    return match task_entity_option {
        Some(_) => StatusCode::NO_CONTENT,
        None => StatusCode::NOT_FOUND,
    };
}

pub async fn delete_task(
    State(task_repository): State<DynTaskRepository>,
    Path(task_id): Path<uuid::Uuid>,
) -> StatusCode {
    let task_entity_option = task_repository
        .lock()
        .await
        .delete_by_task_id(task_id)
        .await;
    return match task_entity_option {
        Some(_) => StatusCode::NO_CONTENT,
        None => StatusCode::NOT_FOUND,
    };
}
